metals/src/main/scala/scala/meta/internal/pantsbuild/BloopPants.scala

```diff
@@ -0,0 +1,657 @@
+    interruptedTry {
+      val cacheDir = Files.createDirectories(
+        args.workspace.resolve(".pants.d").resolve("metals")
+      )
+      val outputFilename = {
+        val processed =
+          args.targets.map(_.replaceAll("[^a-zA-Z0-9]", "")).mkString
+        if (processed.isEmpty()) {
+          MD5.compute(args.targets.mkString) // necessary for targets like "::/"
+        } else {
+          processed
+        }
+      }
+      val outputFile = cacheDir.resolve(s"$outputFilename.json")
+      val bloopDir = Files.createDirectories(args.out.resolve(".bloop"))
+      args.token.checkCanceled()
+
+      val filemap = Filemap.fromPants(args.workspace, args.targets)
+      val fileCount = filemap.fileCount()
+      if (fileCount > args.maxFileCount) {
+        val targetSyntax = args.targets.mkString("'", "' '", "'")
+        scribe.error(
+          s"The target set ${targetSyntax} is too broad, it expands to ${fileCount} source files " +
+            s"when the maximum number of allowed source files is ${args.maxFileCount}. " +
+            s"To fix this problem, configure a smaller set of Pants targets."
+        )
+        throw new CancellationException("too many Pants targets")
+      }
+      args.onFilemap(filemap)
+
+      if (!args.isCache || !Files.isRegularFile(outputFile)) {
+        runPantsExport(args, outputFile)
+      }
+
+      if (Files.isRegularFile(outputFile)) {
+        val text =
+          new String(Files.readAllBytes(outputFile), StandardCharsets.UTF_8)
+        val json = ujson.read(text)
+        new BloopPants(args, bloopDir, json, filemap).run()
+      } else {
+        throw new NoSuchFileException(
+          outputFile.toString(),
+          null,
+          "expected this file to exist after running `./pants export`"
+        )
+      }
+    }
+
+  def bloopRegenerate(
+      workspace: AbsolutePath,
+      targets: List[String]
+  )(implicit ec: ExecutionContext): Unit = {
+    val filemap = Filemap.fromPants(workspace.toNIO, targets)
+    val bloopDir = workspace.resolve(".bloop")
+    for {
+      (target, files) <- filemap.iterator()
+      jsonFile = bloopDir.resolve(BloopPants.makeJsonFilename(target))
+      if jsonFile.isFile
+    } {
+      val json = ujson.read(jsonFile.readText)
+      val newSources =
+        files.iterator.map(file => Value.Str(file.toString)).toBuffer
+      json("project")("sources") = newSources
+      jsonFile.writeText(ujson.write(json, indent = 4))
+    }
+  }
+
+  private def runPantsExport(
+      args: Args,
+      outputFile: Path
+  )(implicit ec: ExecutionContext): Unit = {
+    val pantsBinary = args.workspace.resolve("pants").toString()
+    val command = List[String](
+      pantsBinary,
+      s"--no-quiet",
+      s"--export-libraries-sources",
+      s"--export-output-file=$outputFile",
+      s"export-classpath",
+      s"export"
+    ) ++ args.targets
+    val shortName = "pants export-classpath export"
+    SystemProcess.run(
```

Comment:
Shouldn't we use the process runner from the bloopInstall function?

Reply:
No, since that one launches `metals/slowTask` and we call several system processes in this step.